runtime-sdk/modules/evm/src/signed_query.rs

```diff
@@ -0,0 +1,248 @@
+use std::convert::TryFrom as _;
+
+use ethabi::Token;
+use k256::{ecdsa::recoverable, elliptic_curve::sec1::ToEncodedPoint as _};
+use once_cell::sync::OnceCell;
+use sha3::{Digest as _, Keccak256};
+
+use oasis_runtime_sdk::{
+    context::Context, core::common::crypto::hash::Hash, modules::accounts::API as _,
+};
+
+use crate::{
+    state,
+    types::{Leash, SimulateCallQuery},
+    Config, Error,
+};
+
+pub(super) fn verify<C: Context, Cfg: Config>(
+    ctx: &mut C,
+    query: SimulateCallQuery,
+    mut signature: [u8; 65],
+) -> Result<SimulateCallQuery, Error> {
+    let leash = match query.leash.as_ref() {
+        Some(leash) => leash,
+        None => return Err(Error::InvalidSignedQuery("missing leash")),
+    };
+
+    // First, verify the signature since it's cheap compared to accessing state to verify the leash.
+    let invalid_signature = || Error::InvalidSignedQuery("invalid signature");
+    signature[64] = signature[64].saturating_sub(27); // Some wallets generate high `s`.
+    let sig =
+        recoverable::Signature::try_from(signature.as_slice()).map_err(|_| invalid_signature())?;
+    let signed_message = hash_call_toplevel::<Cfg>(&query);
+    let signer_key = sig
+        .recover_verify_key_from_digest_bytes(&signed_message.into())
+        .map_err(|_| invalid_signature())?;
+    let signer_addr_digest = Keccak256::digest(&signer_key.to_encoded_point(false).as_bytes()[1..]);
+    if &signer_addr_digest[12..] != query.caller.as_ref() {
```

Comment:
Unify this with similar address recovery in `raw_tx`.

Reply:
Okay, but the function is going to return a `[u8;32]` to avoid a copy.

Reply:
<img width="607" alt="Screen Shot 2022-06-10 at 04 01 18" src="https://user-images.githubusercontent.com/2353785/172934066-379287c1-2495-4e02-9abf-afee5c1c5b64.png">